### ISSUE TITLE
Extend JSPONPath with select functions isUndefined, isDefined, isEmpty and isNotEmpty

### ISSUE THREAD

**Issue by MalibuKoKo** [kubemod/kubemod#87]
Hello,

I try understand how select filters mecanism works and what is the syntax we can use inside. Let's me refer to the README: sections [select-string--required](https://github.com/kubemod/kubemod#select-string--required) and [select-filters]( https://github.com/kubemod/kubemod#select-filters).

### How select only containers which did not contains securityContext property or empty object ?
Example:
```
{
    "kind": "Deployment",
    "spec": {
        "template": {
            "spec": {
                "containers": [
                    {
                        "name": "patch-me"
                    },
                    {
                        "name": "me-too",
                        "securityContext": {}
                    },
                    {
                        "name": "not-me",
                        "securityContext": {
                            "runAsNonRoot": true
                        }
                    }
                ]
            }
        }
    }
}
```
I have the impression that the filters are not working as they should be and describe in [JSONPath](https://goessner.net/articles/JsonPath/) documentation.

Indeed, a simple JSONPath likes : **$.spec.template.spec.containers[?(@.name)]** may returns following paths
```
[
  "$['spec']['template']['spec']['containers'][0]",
  "$['spec']['template']['spec']['containers'][1]",
  "$['spec']['template']['spec']['containers'][2]"
]
```
It's not the case with the following ModRule specifications:
```
apiVersion: api.kubemod.io/v1beta1
kind: ModRule
metadata:
  name: kubemod-patch-deployments-containers-securitycontext
  namespace: kubemod-system
spec:
  targetNamespaceRegex: .*
  type: Patch
  match:
    - matchValue: Deployment
      select: $.kind
  patch:
    - op: add
      select: '$.spec.template.spec.containers[?(@.name)]'
      path: '/spec/template/spec/containers/#0/securityContext'
      value: |-
        runAsNonRoot: true
        capabilities:
          drop:
          - ALL
```

I understood that kubemod used the libraries:
- [PaesslerAG/jsonpath](https://github.com/kubemod/kubemod/blob/v0.14.0/go.mod#L10) from this [repository](https://github.com/PaesslerAG/jsonpath)
- [PaesslerAG/gval](https://github.com/kubemod/kubemod/blob/v0.14.0/go.mod#L9) from this [repository](https://github.com/PaesslerAG/gval)

So the response could be in the mecanisms of the [Gval package](https://pkg.go.dev/github.com/PaesslerAG/jsonpath@v0.1.1) ?

**Comment by vassilvk** [comment]
Hi @MalibuKoKo,

You are correct - KubeMod uses PaesslerAG/jsonpath as its internal JSONPath processing library.

Unfortunately, that implementation has a number of documented limitations, including the lack of support for the `[?(@.property)]` presence filter ([see issue here](https://github.com/PaesslerAG/jsonpath/issues/23)).

If you need to perform presence check, PaesslerAG/jsonpath supports syntax `[?(@.property != "")]` ([see this issue](https://github.com/PaesslerAG/jsonpath/issues/8#issuecomment-438679615)).
It's weird that you have to compare with an empty string, but it works for all types of properties, including maps (which is the case with `securityContext`).

So in your example above, `select: '$.spec.template.spec.containers[?(@.securityContext != "")]'` will yield all containers which have `securityContext` field defined. This (again unfortunately for your use case) will include containers where `securityContext` is defined, but points to an empty map.
What makes things worse is that you don't actually need a presence test - you need the opposite - a negative presence test. This is not supported by PaesslerAG/jsonpath.

Looking for alternative implementations of JSONPath, one will come across the [JSONPath implemented in `kubectl`](https://pkg.go.dev/k8s.io/client-go@v0.23.5/util/jsonpath). It supports the `[?(@.property)]` presence test, but it also doesn't support the one you need - a negative presence test - [see this issue](https://github.com/kubernetes/website/issues/7853).

Bottom line, it's a bit of a mess and no popular golang JSONPath implementation seems to support a negative presence test out of the box.

I think it is worth for me to take a look at implementing an extension of the Gval language to support this use case.

**Comment by vassilvk** [comment]
Proposal for extending KubeMod's JSONPath:

## Value `undefined`
- Introduce internal representation of value `undefined`.
- Resolve each path that includes undefined properties to value `undefined`.
  For example `$.a.b.c` will resolve to `undefined` if any of the `a`, `x`, `y` or `z` properties do not exist.
- Filter out all `undefined` values on partial matches.
- Make all equality and arithmetic comparisons to `undefined` return `false`.
  For example, assuming that `$.a.b.c` is a path to undefined property, all of the following expressions will yield `false`:
  - `$.a.b.c == 12`
  - `$.a.b.c != 12`
  - `$.a.b.c > 12`
  -  `$.a.b.c < 12`
  -   `$.a.b.c == true`
  -   `$.a.b.c == false`
- Make all boolean operators (`&&` and `||`) require boolean operands.

## Introduce `undefined`-based functions
- `isDefined()` - returns `true` if the passed in path leads to a defined property, otherwise return `false`.
- `isUndefined()` - returns `true` if the passed in path leads to an undefined property, otherwise return `false`.
- `isEmpty()` - returns `true` if the passed in path leads to one of the following values:
  - An empty array
  - An empty object
  - An empty string
  - Null
  - `undefined`
- `isNotEmpty()` - equivalent to evaluating `!isEmpty()`

## Usage in ModRules

The above extension of JSONPath will enable to solve use cases such as the one raised by @MalibuKoKo.
For example, to patch the containers which have either no `securityContext` defined or `securityContext` is empty, one would use the following KubeMod rule:

```yaml
apiVersion: api.kubemod.io/v1beta1
kind: ModRule
metadata:
  name: kubemod-patch-deployments-containers-securitycontext
  namespace: kubemod-system
spec:
  targetNamespaceRegex: .*
  type: Patch
  match:
    - matchValue: Deployment
      select: $.kind
  patch:
    - op: add
      select: '$.spec.template.spec.containers[? isEmpty(@.securityContex)]'
      path: '/spec/template/spec/containers/#0/securityContext'
      value: |-
        runAsNonRoot: true
        capabilities:
          drop:
          - ALL
```

**Comment by vassilvk** [comment]
Closed by #88

**Comment by MalibuKoKo** [comment]
Hi @vassilvk,

I'm happy to hear that ! It would have been a shame not to be able to patch security context on containers ;-)

Thanks for this implementation, I'm waiting impatiently for the new version to test theses functions.

**Comment by vassilvk** [comment]
Hi @MalibuKoKo,

The v0.15.0-rc-1 is available.

You can install it (or upgrade from v0.14.0) by running the following commands:

```bash
# Delete previous kubemod certificate generation job.
kubectl delete job kubemod-crt-job -n kubemod-system
# Make KubeMod ignore Kubernetes' system namespace.
kubectl label namespace kube-system admission.kubemod.io/ignore=true --overwrite
# Upgrade KubeMod operator.
kubectl apply -f https://raw.githubusercontent.com/kubemod/kubemod/v0.15.0-rc-1/bundle.yaml
```

Please let me know if the new JSONPath processing works as expected.

You should be able to target all containers with either no `securityContext` defined or an empty `securityContext` by using the following `select` in a patch `add` operation (as seen in the example I posted in my proposal above):

```yaml
select: '$.spec.template.spec.containers[? isEmpty(@.securityContext)]'
```
